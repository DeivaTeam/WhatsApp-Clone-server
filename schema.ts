import 'reflect-metadata';
import { AppModule } from './modules/app.module';

const accountsServer: any = { getServices: () => ({ password: {} })};
const connection: any = {};

export default AppModule.forRoot({ accountsServer, connection }).typeDefs;
