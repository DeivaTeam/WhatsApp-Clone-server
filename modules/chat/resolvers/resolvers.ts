import { PubSub, withFilter } from 'apollo-server-express';
import { Chat } from '../models/Chat';
import { ChatProvider } from '../providers/chat.provider';
import { ModuleContext } from '@graphql-modules/core';

export const Subscription = {
  chatAdded: {
    subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('chatAdded'),
      (data: { chatAdded: Chat, creatorId: number }, variables, { injector }: ModuleContext) =>
        data && injector.get(ChatProvider).filterChatAddedOrUpdated(data.chatAdded, data.creatorId)
    ),
  },
  chatUpdated: {
    subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('chatUpdated'),
      (data: { chatUpdated: Chat, updaterId: number }, variables, { injector }: ModuleContext) =>
        data && injector.get(ChatProvider).filterChatAddedOrUpdated(data.chatUpdated, data.updaterId)
    ),
  },
}

