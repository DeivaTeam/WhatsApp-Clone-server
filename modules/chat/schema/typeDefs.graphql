type Query {
  chats: [Chat!]! @auth
  chat(chatId: ID!): Chat @auth
}

type Subscription {
  chatAdded: Chat
  chatUpdated: Chat
}

type Chat {
  #May be a chat or a group
  id: ID!
  createdAt: Date!
  #Computed for chats
  name: String
  #Computed for chats
  picture: String
  #All members, current and past ones. Includes users who still didn't get the chat listed.
  allTimeMembers: [User!]!
  #Whoever gets the chat listed. For groups includes past members who still didn't delete the group. For chats they are the only ones who can send messages.
  listingMembers: [User!]!
  #Actual members of the group. Null for chats. For groups they are the only ones who can send messages. They aren't the only ones who get the group listed.
  actualGroupMembers: [User!]
  #Null for chats
  admins: [User!]
  #If null the group is read-only. Null for chats.
  owner: User
  #Computed property
  isGroup: Boolean!
}

type Mutation {
  addChat(userId: ID!): Chat @auth
  addGroup(userIds: [ID!]!, groupName: String!, groupPicture: String): Chat @auth
  updateGroup(chatId: ID!, groupName: String, groupPicture: String): Chat @auth
  removeChat(chatId: ID!): ID @auth
  addAdmins(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  removeAdmins(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  addMembers(groupId: ID!, userIds: [ID!]!): [ID]! @auth
  removeMembers(groupId: ID!, userIds: [ID!]!): [ID]! @auth
}
