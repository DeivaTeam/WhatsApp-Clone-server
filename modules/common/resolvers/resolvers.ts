import { GraphQLDateTime } from 'graphql-iso-date';
import { ScalarType } from 'graphql-toolkit';

export const Date = GraphQLDateTime;
