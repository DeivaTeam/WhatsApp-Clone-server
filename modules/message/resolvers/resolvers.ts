import { PubSub } from 'apollo-server-express';
import { withFilter } from 'apollo-server-express';
import { Message } from '../models/Message';
import { MessageProvider } from '../providers/message.provider';
import { ModuleContext } from '@graphql-modules/core';

export default {
  Subscription: {
    messageAdded: {
      subscribe: withFilter((root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('messageAdded'),
        (data: { messageAdded: Message }, variables, { injector }: ModuleContext) => data && injector.get(MessageProvider).filterMessageAdded(data.messageAdded)
      ),
    },
  },
};
