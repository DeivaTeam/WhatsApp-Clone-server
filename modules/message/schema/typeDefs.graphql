type Subscription {
  messageAdded: Message
}

enum MessageType {
  LOCATION
  TEXT
  PICTURE
}
