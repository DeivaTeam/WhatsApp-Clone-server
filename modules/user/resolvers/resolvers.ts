import { PubSub, GraphQLUpload } from 'apollo-server-express';
import { withFilter } from 'apollo-server-express';
import { UserProvider } from '../providers/user.provider';
import { User } from '../models/User';
import { ModuleContext } from '@graphql-modules/core';

export const Upload = GraphQLUpload;
export const Subscription = {
  userAdded: {
    subscribe: withFilter(
      (root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('userAdded'),
      (data: { userAdded: User }, variables, { injector }: ModuleContext) => data && injector.get(UserProvider).filterUserAddedOrUpdated(data.userAdded),
    ),
  },
  userUpdated: {
    subscribe: withFilter(
      (root, args, { injector }: ModuleContext) => injector.get(PubSub).asyncIterator('userAdded'),
      (data: { userUpdated: User }, variables, { injector }: ModuleContext) => data && injector.get(UserProvider).filterUserAddedOrUpdated(data.userUpdated)
    ),
  },
}
