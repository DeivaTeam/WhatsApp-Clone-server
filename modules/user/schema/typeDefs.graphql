scalar Upload

type Subscription {
  userAdded: User
  userUpdated: User
}
