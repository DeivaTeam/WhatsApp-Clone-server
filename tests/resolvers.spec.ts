import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server";
import typeDefs from "../schema/typeDefs";
import { resolvers } from "../schema/resolvers";

describe('Apollo Server', () => {
  let server: ApolloServer;

  beforeEach(async() => {
    server = new ApolloServer({
      typeDefs: gql(typeDefs),
      resolvers: resolvers as any,
    });
  });

  it('should match the snapshot for the GetChats query', async () => {
    const {query} = createTestClient(server);

    const res = await query({
      query: gql`
          query GetChats {
              chats {
                  id
                  name
                  messages {
                      id
                      content
                  }
              }
          }
      `,
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });

  it('should match the snapshot for the GetChat query', async () => {
    const {query} = createTestClient(server);

    const res = await query({
      query: gql`
          query GetChat($chatId: ID!) {
              chat(chatId: $chatId) {
                  id
                  name
                  messages {
                      id
                      content
                  }
              }
          }
      `,
      variables: {chatId: 8},
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });

  it('should match the snapshot for the GetUsers query', async () => {
    const {query} = createTestClient(server);

    const res = await query({
      query: gql`
          query GetUsers {
              users {
                  id,
                  name,
                  picture,
              }
          }
      `,
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });

  it('should match the snapshot for the GetMe query', async () => {
    const {query} = createTestClient(server);

    const res = await query({
      query: gql`
          query GetMe {
              me {
                  id,
                  name,
                  picture,
              }
          }
      `,
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
