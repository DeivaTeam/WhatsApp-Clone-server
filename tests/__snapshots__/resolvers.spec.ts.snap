// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apollo Server should match the snapshot for the GetChat query 1`] = `
Object {
  "chat": Object {
    "id": "8",
    "messages": Array [
      Object {
        "content": "I made a group",
        "id": "1",
      },
      Object {
        "content": "Ops, user 3 was not supposed to be here",
        "id": "2",
      },
      Object {
        "content": "Awesome!",
        "id": "3",
      },
    ],
    "name": "A user 0 group",
  },
}
`;

exports[`Apollo Server should match the snapshot for the GetChats query 1`] = `
Object {
  "chats": Array [
    Object {
      "id": "1",
      "messages": Array [
        Object {
          "content": "You on your way?",
          "id": "1",
        },
        Object {
          "content": "Yep!",
          "id": "2",
        },
      ],
      "name": "Avery Stewart",
    },
    Object {
      "id": "2",
      "messages": Array [
        Object {
          "content": "Hey, it's me",
          "id": "1",
        },
      ],
      "name": "Katie Peterson",
    },
    Object {
      "id": "3",
      "messages": Array [
        Object {
          "content": "I should buy a boat",
          "id": "1",
        },
        Object {
          "content": "You still there?",
          "id": "2",
        },
      ],
      "name": "Ray Edwards",
    },
    Object {
      "id": "6",
      "messages": Array [],
      "name": "Niccolò Belli",
    },
    Object {
      "id": "8",
      "messages": Array [
        Object {
          "content": "I made a group",
          "id": "1",
        },
        Object {
          "content": "Ops, user 3 was not supposed to be here",
          "id": "2",
        },
        Object {
          "content": "Awesome!",
          "id": "3",
        },
      ],
      "name": "A user 0 group",
    },
  ],
}
`;

exports[`Apollo Server should match the snapshot for the GetMe query 1`] = `
Object {
  "me": Object {
    "id": "1",
    "name": "Ethan Gonzalez",
    "picture": "https://randomuser.me/api/portraits/thumb/men/1.jpg",
  },
}
`;

exports[`Apollo Server should match the snapshot for the GetUsers query 1`] = `
Object {
  "users": Array [
    Object {
      "id": "2",
      "name": "Bryan Wallace",
      "picture": "https://randomuser.me/api/portraits/thumb/men/2.jpg",
    },
    Object {
      "id": "3",
      "name": "Avery Stewart",
      "picture": "https://randomuser.me/api/portraits/thumb/women/1.jpg",
    },
    Object {
      "id": "4",
      "name": "Katie Peterson",
      "picture": "https://randomuser.me/api/portraits/thumb/women/2.jpg",
    },
    Object {
      "id": "5",
      "name": "Ray Edwards",
      "picture": "https://randomuser.me/api/portraits/thumb/men/3.jpg",
    },
    Object {
      "id": "6",
      "name": "Niccolò Belli",
      "picture": "https://randomuser.me/api/portraits/thumb/men/4.jpg",
    },
    Object {
      "id": "7",
      "name": "Mario Rossi",
      "picture": "https://randomuser.me/api/portraits/thumb/men/5.jpg",
    },
  ],
}
`;
